toplevel/role/approle-creds.go

```diff
@@ -0,0 +1,160 @@
+package role
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/app-sre/vault-manager/pkg/vault"
+	log "github.com/sirupsen/logrus"
+)
+
+func populateApproleCreds(address string, roles []entry, dryRun bool) error {
+	kvVersions, err := getKvEngineVersions(address)
+	if err != nil {
+		return err
+	}
+
+	for _, role := range roles {
+		if strings.ToLower(role.Type) == "approle" && len(role.OutputPath) > 0 {
+			// root of secret path is name of the secret engine
+			pathRoot := strings.Split(role.OutputPath, "/")[0]
+			if _, exists := kvVersions[fmt.Sprint(pathRoot, "/")]; !exists {
+				log.WithFields(log.Fields{
+					"name":     role.Name,
+					"path":     role.OutputPath,
+					"instance": address,
+				}).Info("[Vault Approle] Specified output path does not match any existing KV engines")
+				return errors.New("approle creds invalid output path")
+			}
+
+			// determine if data already exists at desired output path and skip if exists
+			// first determine KV version of desired output
+			var version string
+			switch kvVersions[fmt.Sprint(pathRoot, "/")] {
+			case "1":
+				version = vault.KV_V1
+			case "2":
+				version = vault.KV_V2
+			default:
+				log.WithFields(log.Fields{
+					"name":       role.Name,
+					"path":       role.OutputPath,
+					"kv_version": kvVersions[fmt.Sprint(pathRoot, "/")],
+					"instance":   address,
+				}).Info("[Vault Approle] Retrieved KV version is not supported")
+				return errors.New("approle creds unsupported KV version")
+			}
+			secret, err := vault.ReadSecret(address, role.OutputPath, version)
+			if err != nil {
+				log.WithFields(log.Fields{
+					"name":       role.Name,
+					"path":       role.OutputPath,
+					"kv_version": kvVersions[fmt.Sprint(pathRoot, "/")],
+					"instance":   address,
+				}).Info("[Vault Approle] Unable to read desired output path")
+				return err
```

Comment:
what happens here if the secret does not exist? do we get a `err != nil` , in which case we should rather continue?

Reply:
Ah, I see we'd get `nil, nil`.. sorry